src/rdp.c

```diff
@@ -423,6 +427,20 @@ void rdp_enable_blend_fill( void )
     __rdp_ringbuffer_send();
 }
 
+/**
+ * @brief Enable display of 2D shaded triangles
+ *
+ * This must be called before using #rdp_draw_shaded_triangle.
+ */
+void rdp_enable_shading( void )
+{
+    __rdp_ringbuffer_queue(0x2f002030);
+    __rdp_ringbuffer_queue(0);
+    __rdp_ringbuffer_queue(0x3C0F8E1F);
+    __rdp_ringbuffer_queue(0x001E0904);
+    __rdp_ringbuffer_send();
```

Comment:
For Other Modes, it'll take some looking up. And the others likely know better than I for what's most correct too. For Combine Mode I searched the discord and found I had the same striped triangle result as another and that the value in krom's test did not work on hardware (cen64 does not appear to care). What I ended up with  was simply an attempt to get the output to be just the shade color out of the (A-B)*C+D. IIRC, I tried to get everything else to be zero and put SHADE in D and it worked on hardware so I left it alone after that.

I do not know. For all I know, I've disabled texturing in the Other Modes command or something (which would naturally affect textured rects).
What is probably needed is all the constants/defines for the flags/options in both these commands and an actual function/system to set them, as well as deciding whether it should always be done by the callee or by the actual function checking a cached version and doing what's necessary to change to correct state for their operation. This would be even more necessary as the different triangle and color and blending options permutate (looks like there's a lot!).

Reply:
Also, thank you for the guide, it's really good!

Reply:
I just checked again and the other functions (`rdp_enable_primitive_fill`, `rdp_enable_texture_copy` etc.) all set the Other Modes as well. But instead of `0x2F` they use `0xEF` (I don't know why, I think the former is more correct) which is quite confusing. So if each of these functions represents a state which supports a certain mode of drawing (like "sprites" or "shaded triangles" etc.) it should be fine to set the Other Modes each time. Also AFAIK the Combine Mode has no effect in copy or fill mode, so that should be ok to set here as well.

As long as these values do the job they should be fine. But as I said, to avoid future confusion some documentation is needed. A simple comment should be enough for now.

A new system to generate and queue raw RDP commands is actually something that is planned. We can then just replace these manually assembled values with the new API when it's ready.

Reply:
I think this setup of manually setting the other modes flag will get unwieldy as `rdp.c` grows. Right now functions like `rdp_enable_blend_fill` and `rdp_enable_primitive_fill` will also override the "set other modes" state. It would probably be better to shadow the set other modes bits in some static variable in `rdp.c` and have a separate function that generates the `set other mode` command based on on that static variable. Then in this function you can just set/unset the appropriate bits without effecting other state that people might want to keep such as the sample type, blend state, cvg coverage, etc.

I'm not sure if these changes are appropriate for this PR, or for a future one that will overhaul `rdp.c`. But I agree with @snacchus, make sure you follow the existing conventions in `rdp.c`

Reply:
Actually this will be the most difficult part of the API to decide IMO. There are cases where specific flags should be set in order for a specific primitive to render correctly and it is not trivial to manage. Ideally we would want RDP state commands to be easily composable to be able to set whatever flag the user wants to set but OTOH that will expose too much detail for a high-level API such as we have now in `rdp.c`.

There are a few strategies that we can adopt;

- We can implement a standard media/gl library API but completely covering it will take a long time.
- @rasky's idea is embedding these RDP preparation commands into the primitive calls which will "cache" the state such that succesive calls will not set them over and over. IMO this is a really good idea but interoperability with the lower level commands is a little tricky as the cached versions might get stale if the user issues additional commands and the code may have to peek into all the commands issued, which is something I don't like very much.
- Some form of a compile time composition of commands that can produce command lists that can accept run-time parameters. C preprocessor might be an option but I'm not sure if it will enable enough flexibility (I'm no good at it at all). We might consider another tool or even a custom language implementation. The major drawback would be all calls should be made via that system such that it can decide which settings to re-issue and drop etc. I mean we can statically analyze as mush as we can and optimize beforehand. But this is not a hard requirement, we can use this for caching or use it to implement the gl library. The advantage is that we can change the implementation and even move it to RSP/RDP as we go without breaking existing code much (hopefully). If implementation requires doing command peeks, this will still enable that as all calls go throught the mechanism though we probably don't have to. We can still implement this in a crude way without any optimizations at first. Imagine we create this new method like this (this is totally made up at this point);

    ```
    rdp_draw_shaded_triangle(x1, y1, ..., v1G, v1R...) = rdp_draw_pipe(
    set_other_modes({
        .blend_enable = true,
        ...
    }),
    edge_coeffs(x1, y1,..),
    shade_coeffs(v1G, v1R,..),
    )
    ```

    Then it is possible to compose such constructs into bigger and bigger things;

    ```
    rdp_draw_blended_tri_strip(x1, y1, ..., v1G, v1R...) = rdp_draw_pipe(
    edge_coeffs(x1, y1,..),
    edge_coeffs(x2, y2,..),
    )
    ```
    This is still not a simple thing to achieve but would be really powerful if we have it.

Reply:
First thing: I think that `rdp_enable_shading` is fine they way it is now, with more documentation and maybe a different naming. It follows the existing rdp.c convention and we can't fix that in the context of this PR.

Now thinking about the future. I have a large set of rdp.c changes that I'd like to submit soon, especially to improve support for 2D sprites and custom render modes. Those changes keep the current spirit / architecture of rdp.c and show even better the problem with the current rdp render modes functions, that I will try to summarize.

I think the real culprit is that the RDP command `set other modes` combines render flags in a way that makes it hard to make a composable API. Some flags are in fact related to the actual texture being rendered (eg. color mode, palette), while others are related to render modes which should be specificied at render time (do I want to render opaque or transparent?). On top of this, render modes *do affect* the commands that need to be sent to render; for instance switching from COPY to 1CYCLE is not just a render mode change, but you need to account for it when providing horizontal deltas in sprites.

So basically, if you want to have a "modern API" where you first set a render mode (eg. "I want to do 1CYCLE with this color combiner formula to achieve a cool effect") and then you draw something (eg. "Let me draw this texture which I don't even remember whether it's paletted or 16-bit"), you can't achieve that without having a render mode global state. The code needs to know how to merge the render mode state with the flag bits required to draw the specific texture pixel format, and then later it needs to check the current render mode to understand how to calculates the deltas.

So this is the unfortunate situation we live in. I don't think it's possible to have a powerful 2D (or 3D) API without keeping this state. So, in my local changes of rdp, I'm introducing this concept. There is a global variable where the code stores the current render mode (that is, the 64-bit `set other mode` command that was last sent), and then work on it incrementally when needed, or read it if required to understand how to perform calculations.

This is not a problem with the current rdp.c API, and in fact it works wonderfully. It can become a problem if we want to open up a API to let users enqueue custom RDP commands. If the user sent a `set other modes` command through this custom "enqueue whatever" API, they would effectively introduce a bug with the global cache. So my current solution is *not* to open up such API for now in my commit, and simply add one specific API to configure a custom render mode (which I *know* it will end up as a `set other modes`).

Going even further in the future, I don't think having rdp.c directly using CPU to put commands into RDP is that useful. We briefly discussed it before on Discord, but I think it would be much powerful to have RDP commands go through RSP anyway, and then let RSP grow a few features to speed up 2D games, such as applying the matrix stack to 2D sprite rendering (so that you can have static display lists for sprites and tilemaps and simply move them on the screen using the matrix stack). At that point, the RSP would be the obvious place where to keep the global state of the current render mode; and it would be even possible to add the "enqueue raw RDP commands" API as that would not break anything, as the RSP would see the render mode changes and act accordingly.

Reply:
In the scope of this PR Im ok with a simple mode change function as well even though in the long run that will start to become a problem especially considering our backwards compatibility efforts. I am not opposing to caching the config either, I just dont like that... It might be something we have to have in the end. The composition should ideally enable different implementation strategies including caching - it may even result in better optimization strategies. Also we dont need to map the composable units 1 to 1 to the hardware like in my example, we can come up with a good abstraction only considering the "capabilities" of the sytem. I think this also closely relates to @snacchus's intermediate representation idea. OTOH I havent worked with RDP enough to decide on something.